### PR TITLE
[WIP] Core blocks: Refactor Gallery and Image blocks to avoid RichText warnings

### PR DESCRIPTION
**Pull request by gziolo**
## Description
With #6286 we have deprecated event proxying. 

@youknowriad discovered the following:
> I wonder if we should introduce an explicit onFocus prop, it seems to be something commonly used across blocks even after #5039

This PR tries to fix it.

## How has this been tested?
Make sure that Image and Gallery blocks work like before and there are no warnings on JS console related to `RichText` component.

## Types of changes
Refactoring.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: core-blocks/image/edit.js [hunk 1]
```diff
@@ -389,6 +363,11 @@ class ImageEdit extends Component {
 }
 
 export default compose( [
+	withBlockEditContext( ( { setFocusedElement } ) => {
```
Comment (youknowriad) [inline core-blocks/image/edit.js]: Nice trick, do you think this should be documented or recommended?

Comment (gziolo) [reply core-blocks/image/edit.js]: I did the same for both `BlockEdit` and `Plugin` - it's more work to map context to props, but I like the fact that you don't need to navigate between files to discover how context is injected into a component.

Comment (gziolo) [reply core-blocks/image/edit.js]: See:
https://github.com/WordPress/gutenberg/blob/master/editor/components/block-edit/context.js#L24-L26

Comment (gziolo) [reply core-blocks/image/edit.js]: and https://github.com/WordPress/gutenberg/blob/master/plugins/components/plugin-context/index.js#L17-L19.

Well, we could even create a factory for it :D

File: editor/components/index.js [hunk 2]
```diff
@@ -5,6 +5,7 @@ export { default as AlignmentToolbar } from './alignment-toolbar';
 export { default as BlockAlignmentToolbar } from './block-alignment-toolbar';
 export { default as BlockControls } from './block-controls';
 export { default as BlockEdit } from './block-edit';
+export { withBlockEditContext } from './block-edit/context';
```
Comment (youknowriad) [inline editor/components/index.js]: I'm still on the fence here, what's the difference between making this available to block authors and passing its content as props to `edit` like we do for several other props :)

Comment (youknowriad) [reply editor/components/index.js]: cc @aduth Thoughts?

Comment (gziolo) [reply editor/components/index.js]: We can do `prop drilling` for blocks, too :)
I wasn't happy about exposing this HOC to be honest.

**Comment by aduth** [comment]
Should we just add default `onFocus` prop handling for `RichText` ? The intent with #6286 was more to avoid setting the expectation that we blanket forward TinyMCE events, particularly where we're inconsistent with how those events are forwarded, i.e. argument of `onChange` being the new value, argument of `onFocus` (previously) being the TinyMCE event. But it wasn't to say that we strictly don't want to support any of these props. An `onFocus` seems like a common use-case (obviously we depended on it already), and the implementation here seems particularly complex for plugin authors to recreate on their own.

**Comment by gziolo** [comment]
> Should we just add default `onFocus` prop handling for `RichText` ? The intent with #6286 was more to avoid setting the expectation that we blanket forward TinyMCE events, particularly where we're inconsistent with how those events are forwarded, i.e. argument of `onChange` being the new value, argument of `onFocus` (previously) being the TinyMCE event. But it wasn't to say that we strictly don't want to support any of these props. An `onFocus` seems like a common use-case (obviously we depended on it already), and the implementation here seems quite complex for plugin authors to recreate on their own.

I added a kind of automated default `onFocus` handling for `RichText` in #6419, which uses `setFocusedElement` from `BlockEditContext` to do the trick. We can update this implementation to handle `onFocus` prop, too. Related code:
https://github.com/WordPress/gutenberg/blob/master/editor/components/rich-text/index.js#L201-L205

There is a bigger issue with other components, which this PR unveiled. I'm leaning towards the solution proposed by @youknowriad to pass down `setFocusedElement` method as one of the props injected into `edit` component. It still wouldn't be super easy to use by plugin developers. In this context the only thing we need to implement is to have a way to reseet the state so we could inform the block that focus needs to be removed from `RichText` (when clicking on the image).

File: core-blocks/gallery/gallery-image.js [hunk 3]
```diff
@@ -115,7 +82,13 @@ class GalleryImage extends Component {
 		// Disable reason: Each block can be selected by clicking on it and we should keep the same saved markup
 		/* eslint-disable jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/onclick-has-role, jsx-a11y/click-events-have-key-events */
 		return (
-			<figure className={ className } tabIndex="-1" onKeyDown={ this.onKeyDown } ref={ this.bindContainer }>
+			<figure
+				className={ className }
+				tabIndex="-1"
+				onClick={ this.props.onSelect }
```
Comment (tofumatt) [inline core-blocks/gallery/gallery-image.js]: Drive-by review, but does the change from `onFocus` in the child `RichText ` component to the `onClick` here present any accessibility issues? How does keyboard navigation work after this change (I didn't see it mentioned in the manual testing being done)?

**Review by tofumatt — COMMENTED** [review]
Curious about the a11y of switching to `onClick` but the code seems fine to me...

File: core-blocks/gallery/gallery-image.js (hunk 3 above)
Comment (gziolo) [reply core-blocks/gallery/gallery-image.js]: I haven't tested it with the keyboard, I will double check if it should be updated to `onFocus`.

Comment (gziolo) [reply core-blocks/gallery/gallery-image.js]: Updated to use `onFocus`. It looks like Gallery block isn't navigable with keyboard :(

**Comment by aduth** [comment]
I don't think it's been shown that we need to expose `setFocusedElement` to a block implementer at all. As implemented, the requirement for gallery is that we need some way to reflect that focus has left a `RichText` instance. In the same way we support automatic focus detection, why not should we support automatic blur detection (complications therein notwithstanding)?

**Comment by youknowriad** [comment]
>  why not should we support automatic blur detection (complications therein notwithstanding), rather than putting the impetus on the developer (particularly when they've come to expect onFocus to be managed automatically)?

The issue is that sometimes RichText loses the focus but we still want to keep it selected. (Link modal, toolbar buttons), It's probably doable but not simple because of the fact that popovers show outside the RichText's node.

**Comment by aduth** [comment]
>The issue is that sometimes RichText loses the focus but we still want to keep it selected. (Link modal, toolbar buttons), It's probably doable but not simple because of the fact that popovers show outside the RichText's node.

I'm not sure I understand where we rely on this. Can you detail a simple set of steps to reproduce?

**Comment by gziolo** [comment]
> >The issue is that sometimes RichText loses the focus but we still want to keep it selected. (Link modal, toolbar buttons), It's probably doable but not simple because of the fact that popovers show outside the RichText's node.
> 
> I'm not sure I understand where we rely on this. Can you detail a simple set of steps to reproduce?

Let me update PR so we could discover where it could fail. I think we can make it opt-in/out if necessary.

**Comment by youknowriad** [comment]
> I'm not sure I understand where we rely on this

When you click the "link" button in the RichText toolbar, the RichText loses the focus and it's transfered to the input in the link modal which is rendered outside the RichText which means the isSelected prop of the RichText will move to `false` causing the toolbar and the link modal to disappear.

I did try this a long time ago :)

**Comment by gziolo** [comment]
In my testing, it happens also for all controls inside block's toolbar which makes it impossible to use them :(

**Comment by gziolo** [comment]
Removed `withBlockEditContext` usage with 9da4eb4.

I will play with it a bit more tomorrow. As far as I understand, what we really need is to set `onFocus` on the `BlockEdit` which will reset focused element when focus happens outside of `RichText`.

**Comment by youknowriad** [comment]
We're approaching the 3.0 release and the proxy event handlers will be removed, should we add an explicit `onFocus` prop while we try to come up with a solution here?

**Comment by gziolo** [comment]
Yes, let’s go with explicit onFocus. I didn’t come up with anything that would handle it automatically. As you already noted, this seems like a very difficult use case to solve. I plan to play a bit more with .closest but it probably will fail as it operates on DOM nit React DOM 🙁

**Comment by aduth** [comment]
Note: The explicit `onFocus` minimal implementation was merged in #7029.

**Comment by mcsf** [comment]
@gziolo, given the merge of #7029, what's the status of this one?

**Comment by gziolo** [comment]
@mcsf, the original issue is resolved but there are more issues to solve here as explained in my comment in here https://github.com/WordPress/gutenberg/issues/7463#issuecomment-399378730. Let’s close this PR for now. We can reuse this code later when we have an idea how to improve focus handling for RichText component.
